crates/transaction-generator-lib/src/p2p_transaction_generator.rs

```diff
@@ -7,35 +7,108 @@ use aptos_sdk::{
     transaction_builder::{aptos_stdlib, TransactionFactory},
     types::{chain_id::ChainId, transaction::SignedTransaction, LocalAccount},
 };
-use rand::{
-    distributions::{Distribution, Standard},
-    prelude::SliceRandom,
-    rngs::StdRng,
-    Rng, RngCore, SeedableRng,
-};
+use rand::{distributions::{Distribution, Standard}, prelude::SliceRandom, rngs::StdRng, Rng, RngCore, SeedableRng};
 use std::{cmp::max, sync::Arc};
+use std::cmp::min;
+
+
+/// Specifies how to get a given number of samples from an item pool.
+trait Sampler: Send + Sync {
+    fn sample(&mut self, rng: &mut StdRng, count: usize) -> Vec<usize>;
+}
+
+/// A sampler that samples a random subset of the pool. Samples are replaced immediately.
+struct BasicSampler {
+    pool_size: usize,
+}
+
+impl Sampler for BasicSampler {
+    fn sample(&mut self, rng: &mut StdRng, count: usize) -> Vec<usize> {
+        rand::seq::index::sample(rng, self.pool_size, count).into_vec()
+    }
+}
+
+
+/// A samplers designed for generating a block of P2P transfers without read-write conflicts.
+/// The pool is divided into sub-pools, one of the them being the primary:
+/// it will keep serving the sample requests *without replacement*, until it's depleted.
+/// When the current primary is depleted, another sub-pool takes over and the current one resets.
+struct BurnAndRecycleSampler {
+    /// We store all sub-pools together in 1 Vec: `item_pool[segment_size * x..segment_size * (x+1)]` being the x-th sub-pool.
+    item_pool: Vec<usize>,
+    next_index: usize,
+    segment_size: usize,
+    init_shuffle_done: bool
+}
+
+impl BurnAndRecycleSampler {
+    fn new(num_items: usize, segment_size: usize) -> Self {
+        Self {
+            item_pool: (0..num_items).collect(),
+            next_index: 0,
+            segment_size,
+            init_shuffle_done: false,
```

Comment:
Why do you need this? Can't you just shuffle the item_pool in the constructor?

Reply:
you are right, it's probably more straightforward if we shuffle in the constructor... It's just the constructor then need to take an extra parameter `&mut StdRng` which seems a bit annoying to me.